src/magic.cpp

```diff
@@ -0,0 +1,804 @@
+{
+    experience += nxp;
+}
+
+std::string spell::energy_string() const
+{
+    switch( type->energy_source ) {
+        case hp_energy:
+            return _( "health" );
+        case mana_energy:
+            return _( "mana" );
+        case stamina_energy:
+            return _( "stamina" );
+        case bionic_energy:
+            return _( "bionic power" );
+        default:
+            return "";
+    }
+}
+
+std::string spell::energy_cost_string( const player &p ) const
+{
+    if( energy_source() == none_energy ) {
+        return _( "none" );
+    }
+    if( energy_source() == bionic_energy || energy_source() == mana_energy ) {
+        return colorize( to_string( energy_cost() ), c_light_blue );
+    }
+    if( energy_source() == hp_energy ) {
+        auto pair = get_hp_bar( energy_cost(), p.get_hp_max() / num_hp_parts );
+        return colorize( pair.first, pair.second );
+    }
+    if( energy_source() == stamina_energy ) {
+        auto pair = get_hp_bar( energy_cost(), p.get_stamina_max() );
+        return colorize( pair.first, pair.second );
+    }
+    return "error: energy_type";
+}
+
+std::string spell::energy_cur_string( const player &p ) const
+{
+    if( energy_source() == none_energy ) {
+        return _( "infinite" );
+    }
+    if( energy_source() == bionic_energy ) {
+        return colorize( to_string( p.power_level ), c_light_blue );
+    }
+    if( energy_source() == mana_energy ) {
+        return colorize( to_string( p.available_mana() ), c_light_blue );
+    }
+    if( energy_source() == stamina_energy ) {
+        auto pair = get_hp_bar( p.stamina, p.get_stamina_max() );
+        return colorize( pair.first, pair.second );
+    }
+    if( energy_source() == hp_energy ) {
+        return "";
+    }
+    return "error: energy_type";
+}
+
+bool spell::is_valid() const
+{
+    if( type == nullptr ) {
+        return false;
+    }
+    return type->is_valid();
+}
+
+std::string spell::effect() const
+{
+    return type->effect;
+}
+
+energy_type spell::energy_source() const
+{
+    return type->energy_source;
+}
+
+bool spell::is_valid_target( valid_target t ) const
+{
+    for( valid_target tar : type->valid_targets ) {
+        if( tar == t ) {
+            return true;
+        }
+    }
+    add_msg( m_bad, ( "invalid target" ) );
+    return false;
+}
+
+bool spell::is_valid_target( const tripoint &p ) const
+{
+    bool valid = false;
+    if( Creature *const cr = g->critter_at<Creature>( p ) ) {
+        Creature::Attitude cr_att = cr->attitude_to( g->u );
+        valid = valid || ( cr_att != Creature::A_FRIENDLY && is_valid_target( target_hostile ) ) ||
+                ( cr_att == Creature::A_FRIENDLY && is_valid_target( target_ally ) );
+    }
+    if( p == g->u.pos() ) {
+        valid = valid || is_valid_target( target_self );
+    }
+    return valid || is_valid_target( target_ground );
+}
+
+std::string spell::description() const
+{
+    return _( type->description );
```

Comment:
Does `extract_json_strings.py` already extract the strings needed for this translation to work?

Reply:
i do not know how to answer this question.

Reply:
You can add "spell" [here](https://github.com/CleverRaven/Cataclysm-DDA/blob/664be367e96a4e15238742c82a96946421d19e60/lang/extract_json_strings.py#L159) I think.
Then run `./lang/extract_json_strings.py`.
After that extracted strings should be in the `lang/json/spell_from_json.py`.

Reply:
ok, i've added it to that file, but i don't have python to run it. let me know if i need to obtain it.

Reply:
If you want to test it.
On my side the names and descriptions were extracted successfully.
I don't see other translatable fields in the stells.